os/test/src/PathTests.scala

```diff
@@ -2,16 +2,90 @@ package test.os
+      testSegmentsFromString("""//""", "" :: "" :: "" :: Nil)
+      testSegmentsFromString("""///""", "" :: "" :: "" :: "" :: Nil)
+
+      testSegmentsFromString("""a/""", "a" :: "" :: Nil)
+      testSegmentsFromString("""a//""", "a" :: "" :: "" :: Nil)
+      testSegmentsFromString("""a///""", "a" :: "" :: "" :: "" :: Nil)
+
+      testSegmentsFromString("""ahs/""", "ahs" :: "" :: Nil)
+      testSegmentsFromString("""ahs//""", "ahs" :: "" :: "" :: Nil)
+
+      testSegmentsFromString("""ahs/aa/""", "ahs" :: "aa" :: "" :: Nil)
+      testSegmentsFromString("""ahs/aa//""", "ahs" :: "aa" :: "" :: "" :: Nil)
+
+      testSegmentsFromString("""/a""", "" :: "a" :: Nil)
+      testSegmentsFromString("""//a""", "" :: "" :: "a" :: Nil)
+      testSegmentsFromString("""//a/""", "" :: "" :: "a" :: "" :: Nil)
+    }
+    test("Literals") {
+      test("Basic") {
+        assert(rel / "src" / "Main/.scala" == rel / "src" / "Main" / ".scala")
+        assert(root / "core/src/test" == root / "core" / "src" / "test")
+        assert(root / "core/src/test" == root / "core" / "src/test")
+      }
+      test("Compile errors") {
+
+        compileError("""root / "/" """).check("", nonValidPathSegment(""))
+        compileError("""root / "/ " """).check("", nonValidPathSegment(""))
+        compileError("""root / " /" """).check("", nonValidPathSegment(""))
+        compileError("""root / "//" """).check("", nonValidPathSegment(""))
+
+        compileError("""root / "foo/" """).check("", nonValidPathSegment(""))
+        compileError("""root / "foo//" """).check("", nonValidPathSegment(""))
+
+        compileError("""root / "foo/bar/" """).check("", nonValidPathSegment(""))
+        compileError("""root / "foo/bar//" """).check("", nonValidPathSegment(""))
+
+        compileError("""root / "/foo" """).check("", nonValidPathSegment(""))
+        compileError("""root / "//foo" """).check("", nonValidPathSegment(""))
+
+        compileError("""root / "//foo/" """).check("", nonValidPathSegment(""))
+
+        compileError(""" rel / "src" / "" """).check("", nonValidPathSegment(""))
+        compileError(""" rel / "src" / "." """).check("", nonValidPathSegment("."))
+        compileError(""" rel / "src" / ".." """).check("", nonValidPathSegment(".."))
+
+        compileError { """ root / "src/"  """ }.check("", nonValidPathSegment(""))
+        compileError { """ root / "src/." """ }.check("", nonValidPathSegment("."))
+        compileError { """ root / "src/.." """ }.check("", nonValidPathSegment(".."))
+
+        compileError { """ root / "" """ }.check("", nonValidPathSegment(""))
+        compileError { """ root / "." """ }.check("", nonValidPathSegment("."))
+        compileError { """ root / ".." """ }.check("", nonValidPathSegment(".."))
+
+        compileError(""" root / "hello" / ".." / "world" """).check("", nonValidPathSegment(".."))
+        compileError(""" root / "hello" / "../world" """).check("", nonValidPathSegment(".."))
```

Comment:
Should we allow these scenarios? Allowing literals to generate `ups` seems reasonable, given that they are literal strings, as long as they care canonicalized

Reply:
I think it's ok, it gives lots of flexibility to users.

We should decide what to do with following edge cases:
`root / ".." / foo`
`root / "../.." / foo`
etc.
 
I think for the first one we should fail with compile error saying to use `up` instead of `".."`.
What do you think?

Reply:
Let's allow `root / ".." / foo`. It definitely duplicates `/ os.up /`, but both more concise and arguably more intuitive. And the compile-time-only nature of the conversions should stop the normal problems of untrusted user inputs causing filesystem traversal problems